lmcache/v1/gpu_connector.py

```diff
@@ -361,6 +369,226 @@ def get_shape(self, num_tokens: int) -> torch.Size:
         return torch.Size([kv_size, self.num_layers, num_tokens, self.hidden_dim_size])
 
 
+class VLLMPagedMemGPUConnectorV3(GPUConnectorInterface):
+    def __init__(
+        self,
+        metadata: LMCacheEngineMetadata,
+        device: torch.device,
+        use_gpu: bool = False,
+    ):
+        assert device.type == "cuda", "The device should be CUDA."
+        self.metadata = metadata
+        self.device = device
+        self.use_mla = metadata.use_mla
+        self.chunk_size = metadata.chunk_size
+        self.use_gpu = use_gpu
+        self.kvcaches: Optional[List[torch.Tensor]] = None
+        self.page_buffer_size = 0
+        self.group_layers: Optional[list[int]] = None
+        self.group_hidden_dim_sizes: Optional[list[int]] = None
+        self.group_kv_cache_pointers_on_gpu: Optional[list[torch.Tensor]] = None
+        self.group_tmp_buffer: Optional[list[torch.Tensor]] = None
+
+        self.store_stream = torch.cuda.Stream()
+        self.load_stream = torch.cuda.Stream()
+
+    @classmethod
+    def from_metadata(
+        cls,
+        metadata: LMCacheEngineMetadata,
+        use_gpu: bool = False,
+        device: Optional[torch.device] = None,
+    ) -> "VLLMPagedMemGPUConnectorV3":
+        assert device is not None
+        return cls(metadata, device, use_gpu)
+
+    def init_group_info(self):
+        assert self.metadata.kv_layer_groups_manager.kv_layer_groups
+        self.group_layers = []
+        self.group_hidden_dim_sizes = []
+        if self.use_gpu:
+            self.group_tmp_buffer = []
+        for group in self.metadata.kv_layer_groups_manager.kv_layer_groups:
+            # init layers
+            num_layers = len(group.layer_indices)
+            self.group_layers.append(num_layers)
+
+            # init hidden dim size
+            num_kv_head = 1 if self.metadata.use_mla else group.shape[3]
+            # hidden_dim_size = num_kv_head * head_size
+            hidden_dim_size = num_kv_head * group.shape[-1]
+            self.group_hidden_dim_sizes.append(hidden_dim_size)
+
+            # init tmp buffer
+            if self.use_gpu:
+                tmp_buf_shape = torch.Size(
+                    [
+                        1 if self.use_mla else 2,
+                        num_layers,
+                        self.chunk_size,
+                        hidden_dim_size,
+                    ]
+                )
+                tmp_buf = torch.empty(
+                    tmp_buf_shape, dtype=group.dtype, device=self.device
+                )
+                self.group_tmp_buffer.append(tmp_buf)
```

Comment:
Can we do something to reuse the `get_shapes` and `get_dtypes` function here? Just like

```suggestion
        # outside the loop
        if self.use_gpu:
            dtypes = sellf.metadata.get_dtypes()
            shapes = self.get_shapes(self.chunk_size)
            self.group_tmp_buffer = [
                 torch.empty(shape, dtype=dtype, device=self.device)
                 for shape, dtype in zip(shapes, dtypes)
             ]
```

Reply:
done!